ibpsa/project1-boptest
# Merge NREL's alfalfa code into boptest

@dhblum I think it would be a good idea to stage this in a branch other than master. Consider creating a branch in IBPSA/project1-boptest to merge this into. I do not have write permission.

This commit introduces the concept of separate worker and web server processes. The worker is a distinct docker container that is respnsible for carrying out a simulation. The web server is within a separate container is only responsible for serving http requests. A redis database is used to communicate between the worker and web server. The web server and worker may scale indpendently and it is expected that a deployment would involve many workers to manage multiple simultaneous simulations/tests.

This commit is organized such that it adds to the existing boptest project layout, but it does not disturb any of the existing project layout. The worker depends on and makes extensive use of the existing boptest libraries, especially those APIs in testcase.py as well as the forecast and kpi source code directories.

Currently the web api presented by the web server is conceptually equilvant, but not api compatiable with the existing boptest/master API. A future commit will resolve this difference.

To test this commit:

cd <project-root>
docker-compose build
docker-compose up web worker

After starting the services defined by docker-compose, a web application should be viewable in the browser at http://localhost

* There is a graphql https://graphql.org api viewable at http://localhost/graphql
* This architecture depends on bulk file storage such as Amazon S3. In development mode a mino https://min.io container stands in for S3, and is available at http://localhost:9000. From this interface it is possible to view raw file (test case) uploads and simulation files. The default authentication can be found in the .env file that is included in this commit
* A forecast API has been added, which was not previously available in Alfalfa
